packages/gatsby-cli/src/reporter/index.ts

```diff
@@ -123,49 +168,56 @@ const reporter: Reporter = {
       // chalk was already imported, so we need to retroactively adjust level
       chalk.level = 0
     }
-  },
+  }
   /**
    * Log arguments and exit process with status 1.
    * @param {*} args
    */
-  panic(...args) {
-    const error = reporter.error(...args)
-    trackError(`GENERAL_PANIC`, { error })
-    prematureEnd()
+  panic(errorMeta: ErrorMeta, error?: Error): never {
+    const details = this.error(errorMeta, error)
+    trackError(`GENERAL_PANIC`, { error: details })
+    this.prematureEnd()
     process.exit(1)
-  },
+  }
 
-  panicOnBuild(...args) {
-    const error = reporter.error(...args)
-    trackError(`BUILD_PANIC`, { error })
+  panicOnBuild(errorMeta: ErrorMeta, error?: Error): Error | undefined {
+    const details = this.error(errorMeta, error)
+    trackError(`BUILD_PANIC`, { error: details })
     if (process.env.gatsby_executing_command === `build`) {
-      prematureEnd()
+      this.prematureEnd()
       process.exit(1)
     }
     return error
-  },
+  }
 
-  error(errorMeta, error) {
-    let details = {}
+  error(
+    errorMeta: ErrorMeta,
+    error?: Error | Error[]
+  ): IStructuredError | IStructuredError[] {
+    let details: IErrorDetails = {}
     // Many paths to retain backcompat :scream:
-    if (arguments.length === 2) {
+    if (error) {
       if (Array.isArray(error)) {
-        return error.map(errorItem => this.error(errorMeta, errorItem))
+        return error.map(
+          errorItem => this.error(errorMeta, errorItem) as IStructuredError
+        )
       }
       details.error = error
       details.context = {
-        sourceMessage: errorMeta + ` ` + error.message,
+        sourceMessage: errorMeta + ` ` + error?.message,
       }
-    } else if (arguments.length === 1 && errorMeta instanceof Error) {
+    } else if (errorMeta instanceof Error) {
       details.error = errorMeta
       details.context = {
         sourceMessage: errorMeta.message,
       }
-    } else if (arguments.length === 1 && Array.isArray(errorMeta)) {
+    } else if (Array.isArray(errorMeta)) {
       // when we get an array of messages, call this function once for each error
-      return errorMeta.map(errorItem => this.error(errorItem))
+      return errorMeta.map(
+        errorItem => this.error(errorItem) as IStructuredError
+      )
     } else if (arguments.length === 1 && typeof errorMeta === `object`) {
-      details = Object.assign({}, errorMeta)
+      details = { ...errorMeta }
     } else if (arguments.length === 1 && typeof errorMeta === `string`) {
```

Comment:
can we get rid of these other `arguments.length === 1` things?

Reply:
Ah yes. It's always true anyway, as I've checked for `error` above